File: docs/spec/claude/tools/gen-traceability-index.py
#!/usr/bin/env python3
"""Generate reference/traceability-index.md from the per-document traceability tables.

Usage: python3 docs/spec/claude/tools/gen-traceability-index.py
Run from anywhere; paths are resolved relative to this script.

Definition source: any markdown table row whose first cell is exactly one ID
(e.g. `| INV-SM-1 | statement | impl | evidence |`). Every other occurrence of
the ID anywhere in the tree is recorded as a reference. The script fails loudly
on duplicate definitions and lists IDs that are mentioned but never defined.
"""

import os
import re
import sys
from collections import defaultdict

SPEC_ROOT = os.path.normpath(os.path.join(os.path.dirname(os.path.abspath(__file__)), ".."))
OUT_PATH = os.path.join(SPEC_ROOT, "reference", "traceability-index.md")
ID_RE = re.compile(r"\b(?:INV|REQ)-[A-Z]+-\d+\b")
ROW_RE = re.compile(r"^\|\s*`?((?:INV|REQ)-[A-Z]+-\d+)`?\s*\|(.*)$")

def md_files():
    for root, _, files in os.walk(SPEC_ROOT):
        if os.path.basename(root) == "tools":
            continue
        for f in sorted(files):
            if f.endswith(".md"):
                yield os.path.join(root, f)

def relpath(p):
    return os.path.relpath(p, SPEC_ROOT).replace(os.sep, "/")

def strip_md(cell):
    cell = re.sub(r"\[([^\]]*)\]\([^)]*\)", r"\1", cell)  # links -> text
    cell = cell.replace("`", "").replace("**", "").strip()
    return re.sub(r"\s+", " ", cell)

definitions = {}          # id -> (relpath, statement)
duplicates = []           # (id, first_path, dup_path)
mentions = defaultdict(set)  # id -> set of relpaths

index_rel = "reference/traceability-index.md"

for path in md_files():
    rp = relpath(path)
    if rp == index_rel:
        continue
    with open(path, encoding="utf-8") as fh:
        text = fh.read()
    for ident in ID_RE.findall(text):
        mentions[ident].add(rp)
    for line in text.splitlines():
        m = ROW_RE.match(line.strip())
        if not m:
            continue
        ident = m.group(1)
        cells = [c.strip() for c in m.group(2).split("|")]
        statement = strip_md(cells[0]) if cells else ""
        if ident in definitions and definitions[ident][0] != rp:
            duplicates.append((ident, definitions[ident][0], rp))
        elif ident not in definitions:
            definitions[ident] = (rp, statement)

undefined = sorted(i for i in mentions if i not in definitions)

def sort_key(ident):
    kind, area, num = ident.split("-")
    return (area, kind, int(num))

lines = []
lines.append("# Traceability Index")
lines.append("")
lines.append("> **Status:** Generated. Do not edit by hand — regenerate with")
lines.append("> `python3 docs/spec/claude/tools/gen-traceability-index.py` after changing any")
lines.append("> traceability table. ID scheme: [governance.md §2](../governance.md#traceability).")
lines.append("")
lines.append("Single collection point for every `INV-*` / `REQ-*` ID in the specification tree.")
lines.append("**Defined in** is the document whose traceability table owns the ID (statement,")
lines.append("implementation, and verification evidence live there). **Referenced in** lists every")
lines.append("other document that mentions the ID.")
lines.append("")

by_area = defaultdict(list)
for ident in sorted(definitions, key=sort_key):
    by_area[ident.split("-")[1]].append(ident)

total = len(definitions)
lines.append(f"{total} IDs across {len(by_area)} areas.")
lines.append("")

for area in sorted(by_area):
    lines.append(f"## {area}")
    lines.append("")
    lines.append("| ID | Statement | Defined in | Referenced in |")
    lines.append("| --- | --- | --- | --- |")
    for ident in by_area[area]:
        def_path, statement = definitions[ident]
        refs = sorted(mentions[ident] - {def_path})
        ref_links = ", ".join(f"[{r}](../{r})" for r in refs) if refs else "—"
        if len(statement) > 160:
            statement = statement[:157] + "..."
        statement = statement.replace("|", "\\|")
        lines.append(f"| `{ident}` | {statement} | [{def_path}](../{def_path}) | {ref_links} |")
    lines.append("")

if undefined:
    lines.append("## Mentioned but not defined")
    lines.append("")
    lines.append("IDs used somewhere in the tree without a defining traceability-table row —")
    lines.append("each needs a definition or the mention removed:")
    lines.append("")
    for ident in undefined:
        where = ", ".join(f"[{r}](../{r})" for r in sorted(mentions[ident]))
        lines.append(f"- `{ident}` — {where}")
    lines.append("")

os.makedirs(os.path.dirname(OUT_PATH), exist_ok=True)
with open(OUT_PATH, "w", encoding="utf-8") as fh:
    fh.write("\n".join(lines) + "\n")

print(f"wrote {relpath(OUT_PATH)}: {total} IDs, {len(undefined)} undefined mentions")
for ident, first, dup in duplicates:
    print(f"DUPLICATE definition: {ident} in {first} and {dup}", file=sys.stderr)
if duplicates:
    sys.exit(1)

File: docs/spec/claude/tools/gen-traceability-index.js
#!/usr/bin/env node
"use strict";

const fs = require("node:fs");
const path = require("node:path");
const {
    ID_GLOBAL_RE,
    ID_PATTERN,
    SPEC_ROOT,
    markdownFiles,
    readText,
    specRelative,
    stripMarkdown
} = require("./shared/traceability-utils");

const OUT_PATH = path.join(SPEC_ROOT, "generated", "traceability-index.md");
const ROW_RE = new RegExp(`^\\|\\s*\`?(${ID_PATTERN})\`?\\s*\\|(.*)$`);
const definitions = new Map();
const duplicates = [];
const mentions = new Map();

for (const document of markdownFiles()) {
    const relative = specRelative(document);
    const markdown = readText(document);
    for (const ident of markdown.match(ID_GLOBAL_RE) || []) {
        if (!mentions.has(ident)) mentions.set(ident, new Set());
        mentions.get(ident).add(relative);
    }
    for (const line of markdown.split(/\r?\n/)) {
        const match = line.trim().match(ROW_RE);
        if (!match) continue;
        const cells = match[2].split("|").map((cell) => cell.trim());
        if (cells.length < 4) continue;
        const ident = match[1];
        const state = stripMarkdown(cells[0]);
        const statement = stripMarkdown(cells[1]);
        if (
            definitions.has(ident) &&
            definitions.get(ident).document !== relative
        ) {
            duplicates.push([ident, definitions.get(ident).document, relative]);
        } else if (!definitions.has(ident)) {
            definitions.set(ident, { document: relative, state, statement });
        }
    }
}

const undefinedIds = [...mentions.keys()]
    .filter((ident) => !definitions.has(ident))
    .sort();
const sortKey = (ident) => {
    const [kind, area, number] = ident.split("-");
    return [area, kind, Number(number)];
};
const compareIds = (left, right) => {
    const a = sortKey(left);
    const b = sortKey(right);
    return a[0].localeCompare(b[0]) || a[1].localeCompare(b[1]) || a[2] - b[2];
};
const byArea = new Map();
for (const ident of [...definitions.keys()].sort(compareIds)) {
    const area = ident.split("-")[1];
    if (!byArea.has(area)) byArea.set(area, []);
    byArea.get(area).push(ident);
}

const lines = [
    "# Traceability Index",
    "",
    "> **Status:** Generated. Do not edit by hand — regenerate with",
    "> `yarn spec:refresh` after changing any traceability table.",
    "> ID scheme: [governance.md §2](../governance.md#traceability).",
    "",
    "Single collection point for every `INV-*` / `REQ-*` ID in the specification tree.",
    "**Defined in** is the document whose traceability table owns the ID (lifecycle state,",
    "statement, implementation, and verification evidence live there). **Referenced in** lists every",
    "other document that mentions the ID.",
    "",
    `${definitions.size} IDs across ${byArea.size} areas.`,
    ""
];

for (const area of [...byArea.keys()].sort()) {
    lines.push(
        `## ${area}`,
        "",
        "| ID | State | Statement | Defined in | Referenced in |",
        "| --- | --- | --- | --- | --- |"
    );
    for (const ident of byArea.get(area)) {
        const definition = definitions.get(ident);
        const references = [...(mentions.get(ident) || [])]
            .filter((reference) => reference !== definition.document)
            .sort();
        const referenceLinks = references.length
            ? references
                  .map((reference) => `[${reference}](../${reference})`)
                  .join(", ")
            : "—";
        let statement = definition.statement;
        if (statement.length > 160) statement = statement.slice(0, 157) + "...";
        statement = statement.replaceAll("|", "\\|");
        lines.push(
            `| \`${ident}\` | ${definition.state} | ${statement} | [${definition.document}](../${definition.document}) | ${referenceLinks} |`
        );
    }
    lines.push("");
}

if (undefinedIds.length) {
    lines.push(
        "## Mentioned but not defined",
        "",
        "IDs used somewhere in the tree without a defining traceability-table row —",
        "each needs a definition or the mention removed:",
        ""
    );
    for (const ident of undefinedIds) {
        const locations = [...mentions.get(ident)]
            .sort()
            .map((reference) => `[${reference}](../${reference})`)
            .join(", ");
        lines.push(`- \`${ident}\` — ${locations}`);
    }
    lines.push("");
}

fs.mkdirSync(path.dirname(OUT_PATH), { recursive: true });
fs.writeFileSync(OUT_PATH, lines.join("\n") + "\n");
process.stdout.write(
    `wrote ${specRelative(OUT_PATH)}: ${definitions.size} IDs, ${undefinedIds.length} undefined mentions\n`
);
for (const [ident, first, duplicate] of duplicates) {
    process.stderr.write(
        `DUPLICATE definition: ${ident} in ${first} and ${duplicate}\n`
    );
}
if (duplicates.length) process.exit(1);
